### PR TITLE
feat(stepfunctions-tasks): add cpu and memory parameters to EcsRunTask

### LINKED ISSUE #30027: aws-stepfunctions-tasks: Add cpu and memory parameters to EcsRunTask
### Describe the feature

Add cpu and memory parameters to EcsRunTask in StepFunctions to allow flexible allocation of Ecs resources according to the situation.

### Use Case

This is useful when you want to allocate resources flexibly according to the data set used in EcsTask, or when you want to increase memory when retrying from errors such as OOM.

### Proposed Solution

Add cpu and memory parameters to EcsRunTask

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.139.1

### Environment details (OS name and version, etc.)

client:  macOS,Windows,Linux

### PR DESCRIPTION
**Pull request by kackyt**
### Issue # (if applicable)

Closes #30027 .

### Reason for this change

As described in the issue.

### Description of changes

Add cpu and memoryMiB property to EcsRunTaskProps.

### Description of how you validated changes

Add unit tests.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by kackyt** [comment]
Exemption Request : Integration testing is not required as the changes are minor.
> Features must contain a change to an integration test file and the resulting snapshot.

**Comment by kackyt** [comment]
Review please!

**Comment by pahud** [comment]
Hi

This PR has been pending for community review for a while. Please consider posting it on the #contributing channel in cdk.dev. Community members will be on the lookout there as well for possible reviews.

Check [How to get your P2 PR community reviewed](https://github.com/aws/aws-cdk/wiki/CDK-Community-PR-Reviews#how-to-get-your-p2-pr-community-reviewed) for more details.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts
Comment (tmokmss) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:100]: Why not use number type to be consistent with [other similar props](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_stepfunctions_tasks.ContainerOverride.html#cpu)?

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts
Comment (tmokmss) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:453]: Given that there are [many more props](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_TaskOverride.html) to be overriden, shouldn't we use an interface to pass these arguments instead of positional arguments?

e.g. 
```ts
function renderOverrides(props: OverrideProps) 

interface OverrideProps {
 cpu?: number;
 memoriMib?: number;
 containerOverrides?: ContainerOverride[];
}
```

**Review by tmokmss — COMMENTED** [review]
Thanks for the PR! I left some comments but otherwise LGTM.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts
Comment (kackyt) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:100]: @tmokmss 
In the AWS API and task definitions, cpu is string type, and changing it to number type may cause unexpected behavior.

https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-ecs/lib/base/task-definition.ts#L157

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts
Comment (kackyt) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:453]: Thank you. I'll fix it.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts
Comment (tmokmss) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:100]: > may cause unexpected behavior.
Not really, we can just stringfy it after receiving the value. All the existing code will remain unchanged.

Then, the problem is whether we should use string or number. I'm not sure why the current code is inconsistent (e.g. TaskDefinition uses string, while FargateTaskDefinition uses number). Maybe looking through git history will help to find why. If there is no particular reason, I'd prefer number type because the existing ContainerOverride interface also uses number, which is closer to this interface than TaskDefinition construct.

Comment (kackyt) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:100]: All right. I'll fix ix.

Comment (kackyt) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:100]: @tmokmss 
I tried to fix it, but simply changing to number type failed the integration test. 
```
Failed resources:
aws-sfn-tasks-ecs-fargate-run-task | 5:48:49 PM | UPDATE_FAILED        | AWS::StepFunctions::StateMachine      | StateMachine (StateMachine2E01A3A5) Resource handler returned message: "Invalid State Machine Definition: 'SCHEMA_VALIDATION_FAILED: The value for the field 'Cpu' must be a STRING at /States/FargateTask/Parameters' (Service: AWSStepFunctions; Status Code: 400; Error Code: InvalidDefinition; Request ID: ed6dadc7-f05a-474e-abfa-bc5291934ab8; Proxy: null)" (RequestToken: e1181314-190c-640a-43ca-ec6ccd123530, HandlerErrorCode: InvalidRequest)
```

Where in the code do you think we should stringfy this value?

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts
Comment (kackyt) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts]: done 8bb2d45

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts
Comment (kackyt) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts]: done 7061480

Comment (tmokmss) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts]: renderOverrides is where CFn template is actually rendered, so you can stringify the values there. To stringify the value, I'd recommend to look at the existing code such as FargateContainerDefinition. You have to take care of CDK tokens.

Comment (tmokmss) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts]: Sorry, I now find that they are already inconsistent at CFn level... [ContainerOverride](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerOverride.html) uses integer as cpu while [TaskOverride](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_TaskOverride.html) uses string as cpu. Given that, I think it's better to just follow CFn type definitions, because in the future CFn may start to use an actual string that cannot be parsed to a number. What do you think? The below API is so weird but we won't override two cpu props at once anyway...

```ts
const runTask = new tasks.EcsRunTask(this, 'Run', {
  // ...
  cpu: '1024',
  containerOverrides: [
    cpu: 1024,
  ]
});
```

Comment (tmokmss) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts]: We cannot change the existing type definition, so we have to choose which one is better:

```ts
// option 1: weird but maybe safer
const runTask = new tasks.EcsRunTask(this, 'Run', {
  // ...
  cpu: '1024',
  containerOverrides: [
    cpu: 1024,
  ]
});

// option 2: more natural, but prone to future breaking changes
const runTask = new tasks.EcsRunTask(this, 'Run', {
  // ...
  cpu: 1024,
  containerOverrides: [
    cpu: 1024,
  ]
});
```

I now prefer option 1, but anyway I'll approve the review once the all the PR check passes, and let the core team decide it. Any thoughts?

Comment (kackyt) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:100]: Done (option 1) [8c79da9](https://github.com/aws/aws-cdk/pull/30140/commits/8c79da968b328cf4429668afccfd5ed416edbb36)

**Review by tmokmss — APPROVED** [review]
To the maintaners: can you confirm the decision made [here](https://github.com/aws/aws-cdk/pull/30140#discussion_r1705206227)? Or let us know if there are other options 🙏

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts
Comment (shikha372) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:472]: can we keep this same as before `!containerOverrides || containerOverrides.length === 0`,  it was more readable than current negation of negation.

**Comment by shikha372** [comment]
Thanks @kackyt for submitting the PR, want to be double sure that we're not adding the functionality that can already be achieved with the help of container override. Could you help in providing a specific use case of ECS run task where this was not working for you. If you've a code snippet that you're using that would help too.

**Comment by kackyt** [comment]
@shikha372 

This is useful when you want to allocate resources flexibly according to the data set used in EcsTask, or when you want to increase memory when retrying from errors such as OOM.Furthermore, in Fargate Task, this parameter is required to specify the size. We cannot change the size of the Fargate with container override.

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts
Comment (kackyt) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:472]: @shikha372 
I am sorry. However, it is a complex branch and we do not want to clone the same expression. Any suggestions on how to write better code would be appreciated.

Comment (shikha372) [reply packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:451]: Thanks @kackyt , looked at it my initial thought was it would be `!containerOverrides || containerOverrides.length === 0` but seems like not, it would be something like `if (containerOverrides || containerOverrides.length !== 0)` in which case its a negation anyways, i would go with the current one.

**Comment by shikha372** [comment]
Thanks @kackyt, checked the differences with the deployment, i'm not sure i completely agree with statement that it is not supported in Fargate Cluster, I see this getting populated in the state machine definition for both Ec2 and Fargate, let me know if i'm missing something here. Here is the code snippet: 

```
        "TaskDefinition": "ECREc2StackTD10AFFC0F",
        "Overrides": {
          "ContainerOverrides": [
            {
              "Name": "TheContainer",
              "Cpu": 512,
              "Memory": 1024,
              "Environment": [
                {
                  "Name": "SOME_KEY",
                  "Value.$": "$.SomeKey"
                }
              ]
            }
          ]
        },
        "LaunchType": "EC2"
```

```
        "Overrides": {
          "ContainerOverrides": [
            {
              "Name": "TheContainer",
              "Cpu": 512,
              "Memory": 1024,
              "Environment": [
                {
                  "Name": "SOME_KEY",
                  "Value.$": "$.SomeKey"
                }
              ]
            }
          ]
        },
        "LaunchType": "FARGATE",
        "PlatformVersion": "1.4.0"
```
That being said, adding these values help to override the cpu and memory set at task definition level which i don't think exist today, and would review PR for that requirement.

Task override: 
```
        "Overrides": {
          "Cpu": "1024",
          "Memory": "2048",
          "ContainerOverrides": [
            {
              "Name": "TheContainer",
              "Environment": [
                {
                  "Name": "SOME_KEY",
                  "Value.$": "$.SomeKey"
                }
              ]
            }
          ]
        },
        "LaunchType": "FARGATE",
```

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts
Comment (shikha372) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:96]:
```suggestion
   * @see https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_TaskOverride.html
```

File: packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts
Comment (shikha372) [inline packages/aws-cdk-lib/aws-stepfunctions-tasks/lib/ecs/run-task.ts:103]:
```suggestion
   * @see https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_TaskOverride.html
```

**Review by shikha372 — APPROVED** [review]

**Pull request merged**
